mockbroker_test.go

```diff
@@ -47,80 +72,134 @@ func (b *mockBroker) Addr() string {
 }
 
 func (b *mockBroker) Close() {
+	close(b.expectations)
 	if len(b.expectations) > 0 {
-		b.t.Errorf("Not all expectations were satisfied in mockBroker with ID=%d! Still waiting on %d", b.BrokerID(), len(b.expectations))
+		buf := bytes.NewBufferString(fmt.Sprintf("mockbroker/%d: not all expectations were satisfied! Still waiting on:\n", b.BrokerID()))
+		for e := range b.expectations {
+			_, _ = buf.WriteString(spew.Sdump(e))
+		}
+		b.t.Error(buf.String())
 	}
-	close(b.expectations)
+	close(b.closing)
 	<-b.stopper
 }
 
-func (b *mockBroker) serverLoop() (ok bool) {
-	var (
-		err  error
-		conn net.Conn
-	)
-
+func (b *mockBroker) serverLoop() {
 	defer close(b.stopper)
-	if conn, err = b.listener.Accept(); err != nil {
-		return b.serverError(err, conn)
+	var err error
+	var conn net.Conn
+
+	go func() {
+		<-b.closing
+		safeClose(b.t, b.listener)
+	}()
+
+	wg := &sync.WaitGroup{}
+	for i := 0; ; i++ {
+		if conn, err = b.listener.Accept(); err != nil {
+			goto done
```

Comment:
in fact, you can use the `Accept` call in the for loop itself and track the index separately, and then you don't need the additional flow control at all